<a href="https://colab.research.google.com/github/marklide/Nbody/blob/master/Intro_to_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Python3


## Basics

I'd encourage people to experiment with the code snippets. If something seems interesting, run it and see what it does.

Hello World - Traditionally the first program people write when learning a new language.

In [ ]:
print("Hello World!")

Hello World!


print() is a function that takes its arguments and outputs them to the console. In Colab, the console is the white box underneath the code where you can see the result of your code and later also interact with the code.  
print takes any number of arguments that may be of any type. They will be separated by a space.  
So we could also write:

In [ ]:
print("Hello", "World!")

Hello World!


But we are here to use coding for math.   
Simple expressions are valid code, like:

In [ ]:
1+1

2

This also works the same for -, /, and &ast;. However, power is &ast;&ast; instead of ^.  
If we do write 5 ^ 2, python won't complain or crash but the result will be something very different. ^ in python represents something called bitwise xor, but that's not something you're probably ever going to need so just never use ^ to not get confused.  
Note: Those coming from Scala and most other programming languages might expect / to do integer division, e.g. 5/2 is 2, but in python3 the result of division is always a float, so 5/2 will be 2.5. The // operator can instead be used for integer division and it always rounds towards negative infinity, e.g. -5//2 is -3.

In [ ]:
5 - 2

3

In [ ]:
5 / 2

2.5

In [ ]:
5//2

2

In [ ]:
5 * 2

10

Parentheses can also be used. However, note that all multiplication must be explicit, e.g. 3(1 + 2) does not work. Instead write: 3 * (1 + 2).

In [ ]:
4 * (5 - 2) / 5

2.4

There is also the `%` operator, which calculates the modulo or remainder.  
For example, if we did a division and expressed our answer as a mixed number like $\frac{7}{2} = 3\frac{1}{2}$, then the 1 is the remainder.

In [ ]:
7 % 2

1

In [ ]:
9 % 3

0

The operators `+` and `*` also work on text. Plus appends text, while multiplication repeats it.

In [ ]:
'Hello' + ' ' + 'World!'

'Hello World!'

In [ ]:
'Hello World!' * 3

'Hello World!Hello World!Hello World!'

All data in python belongs to a type and that affects what can be done with it and what it can store.  
The three we've already seen are:
- string (text)
- int (integers)
- float (decimal numbers)

Everything in python belongs to some type, and that type can be determined with the type() function.

- strings are always surrounded by two ' or ", like 'Some text'
- floats are numbers with a decimal or scientific notation

In [ ]:
type(3.14)

float

In [ ]:
type('The answer to life, the universe, and everything')

str

In [ ]:
type(42)

int

In [ ]:
"This is a piece of text"

'This is a piece of text'

In [ ]:
0.00000000012

1.2e-10

In [ ]:
1.2e-10

1.2e-10

In [ ]:
42

42

We can also convert between types with int(), float(), and str()
- int() converts to an integer. For float, it will truncate the decimal part
- float() converts to a floating-point number
- str() converts to a string

In [ ]:
int(3.9)

3

In [ ]:
int("42")

42

In [ ]:
float(5)

5.0

In [ ]:
float("3.14")

3.14

Because "3.14" is surrounded by quotation marks, it is stored as a string.  
Note: As long as it is stored as a string, math cannot be done with it. So if you want to do math, use `float()` on it first like in the example above.

In [ ]:
str(5)

'5'

In [ ]:
str(3.14)

'3.14'

If we try to convert values that can't be converted, we'll get an error.

In [ ]:
float("Hello World")

ValueError: ignored

For string to float or int conversion to not error, the string must contain only a constant of the type we're trying to convert to.  
For example, `3.14` is a valid float, `6.02e23` is also a valid float. Thus `float()` could convert these to the `float` type without issue.  
However, if we run `int("3.14")`, it fails because `3.14` is not an integer constant, but a decimal number.

In [ ]:
int("3.14")

ValueError: ignored

If we wanted to get `3` from `"3.14"`, then we can use `float()` to turn it from a string into a `float`, and then use `int()` to convert from `float` to `int`.  
This works because `float()` can parse the string and `int()` can truncate a `float` to `int`.

In [ ]:
int(float("3.14"))

3

A quick but important note on float:  
Floating-point numbers are not precise. For example, certain numbers like 0.1 and 0.2 cannot be fully stored in binary as in binary they are repeating, like 1/3 in base 10 is 0.333... So if we add decimals like 0.1 and 0.2, we can get rounding errors, like in the example below.

In [ ]:
0.1 + 0.2

0.30000000000000004

Issues like the one above are not unique to python, they are near-universal to computers just because of how floating-point numbers are represented in binary and how arithmetic is done on them by the hardware. Beware of this especially when checking if some value equals another float as will be covered later, as otherwise, we might get strange results.

### Comments

Sometimes it is helpful to add a comment to our code to explain our reasoning or to remember some important detail.

In [ ]:
3 ** 5 # 3 to the power 5

A single-line comment starts with `#` and goes to the end of the line. It is ignored by the python interpreter.  
Multiline comments can be made by surrounding the text to comment out with `'''`. This is technically a string, so multiline strings can also be added to code with this.

In [ ]:
'''
Anything here is considered text and not code
3 + 5
The line above is not computed as it is considered text
'''
3 / 5

0.6

## Variables

If we could only add and subtract constants, we would have nothing more than a basic calculator. So we have variables that store data so that we can write code that can calculate a result for many inputs.

In [ ]:
x = 5

Variable assignment in python is easier than in most languages as there are no keywords or other specifiers required. We can then use x to refer to the value 5 that we set it to (as above). Note that the following command does not change the value of x, so if this line is run twice, it will still output 7.

In [ ]:
x + 2

7

Note: x is not a constant, and x = 5 is not an equality in the mathematical sense but rather an assignment.  
Thus a statement like the one below makes sense in python. This does modify x, so if it is run again, the output will be one higher.

In [ ]:
x = x + 1
x

6

Near the beginning I mentioned that we can't have implied multiplication like `4(1+1)`, but instead need to write `4*(1+1)`.  
Likewise, with variables multiplication must be explicit so `4x` is a SyntaxError, but `4 * x` is not.  

In [ ]:
7 * x

42

When there is an assignment of a variable, it is always the variable on the left that is being set to the value of the expression on the right. So in the following line of code, y is being set to the value of x.

In [ ]:
y = x
y

6

The previous example may look similar to a relation or function in the mathematical sense, but if x changes, y will not until we run that piece of code again. Similarly, writing x = t + 1 will fail because t hasn't been given a value yet.  
In this case, what we are looking for are functions.

## Functions

Let's say we want to be able to calculate the y value for every x in a parabola.
In mathematics, this could be expressed as  
f(x) = x<sup>2</sup>  
In python, we can get something similar using functions

In [ ]:
def f(x):
  return x ** 2

Every function starts with the `def` keyword, followed by the function name, and its parameters in parentheses. Parameters are the values that the function will take in. Think of the parameters like variables that are specified when you call the function. For example, if we say `f(5)`, then `x ** 2` is going to be run with x being `5`.     
After the parentheses is a colon, which denotes the start of the function's code.  
All code within the function must be indented. This indentation must be consistent throughout the code. So use either only spaces or only tabs, and always use the same amount.  
To specify what value the function will return, use the `return` keyword followed by whatever data we want to output. 

To run the function, write the name with parentheses next to it with the arguments, like so:

In [ ]:
f(5)

25

When `f(5)` is called, the code in the function is run with x being set to 5.  
Note: Evaluating a function by executing `f(5)` is called calling the function.  
Note: We never put a constraint on what type x can be, so we could run `f('text')` for example, but that will crash because squaring a string doesn't make sense either to us or python.

On terminology,

```
def f(x):
  return x ** 2

f(5)
```

- `x` is the parameter. It is specified when running the function.
- `5` is the argument. It specifies what `x` (the parameter) is going to be.
- `f` is the name of the function.
- `f(5)` is called calling the function. However, I will also use running or executing.
- `25` is the value returned by the function. It is what the expression after the `return` statement evaluates to (`5**2` here because we ran `f` with an argument of `5`). 


Functions can also have multiple inputs, so if we want to have a 3D parabola, we could write:

In [ ]:
def g(x, y):
  return x ** 2 + y ** 2

Functions in Python can do a lot more than mathematical functions  
Note: The variables used within a function are local to a function, so although we use x and y in our function, they aren't affected by our function. This is called scope, in that all the variables used within a scope are only part of that scope and don't exist anymore when the scope is left (when the function's code is complete).

In [ ]:
x = 4

def setX():
  x = 8

print(x)

4


Here, the x outside the function and the x inside of the function are two separate variables, and the one inside of the function stops existing after the function is done running.  
Functions can, however, read from variables outside, they just can't write to them.

In [ ]:
def printX():
  print(x)

printX()

4


Note: This function does not return the value of x. It may seem the same as a function that says:

```
def returnX():
  return x
```

However, `print()` sends data for us to see. Anything sent to `print()` is only useful for us, not the program. So if we wanted to get the value of x in our program, we would use `returnX()`.



When naming functions, it is best to describe what the function does in the name. So try to avoid using f and g as I have done

In [ ]:
def square(x):
  return x ** 2

In [ ]:
def average(a, b, c, d):
  return (a + b + c + d) / 4

average(1,2,3,4)

2.5

Functions can also call other functions like:

In [ ]:
def distance(x, y):
  return ( square(x) + square(y) ) ** 0.5


distance(1, 1)

1.4142135623730951

Functions in Python can be passed as arguments and assigned to variables just like integers or strings.

In [ ]:
def foo():
  print('foo has been called')

def caller(func):
  print('calling function')
  func()

caller(foo)

calling function
foo has been called


`foo` prints `'foo has been called'` to the output when run.  
  
`caller` will first print `'calling function'`, and then call whatever function it was given.  
  
So when `caller(foo)` is called, the parameter `func` is set to `foo` and then `print('calling function')` runs. After that, `func()` is run, but because `func` is set to `foo`, `foo` is called.

In [ ]:
print('setting foo to bar')
bar = foo

bar()

setting foo to bar
foo has been called


In the example above we are setting the variable `bar` to be the function `foo`. Note that this does not change the function so even though we call `bar()` it prints `foo has been called`.

All the functions I've presented use the `return` keyword, but that is not necessary. When a function exits without a return statement, it will return `None`. Think of `None` like a constant used whenever no value is available.

Note for Scala/Java/C++ programmers: `None` is unlike null. `None` is an object and has its own type, `NoneType`. 

In [ ]:
def thisFunctionReturnsNone():
  pass

Note: `pass` is a keyword used when no action is desired. Syntactically at least one statement is always required in a block of code like that of a function.

There is also a collection of built-in functions. `print()` is one we have seen already, and it prints its arguments to the screen and returns nothing, or `None` to be more specific. There is another function that does the opposite of `print`, `input()`. While `print` takes strings and displays them on the screen, `input` asks for input from the user and returns it as a string. To let the user know what they are supposed to enter, `input` also, like `print`, will write its arguments to the screen.

In [ ]:
userInput = input('Enter some text: ')
print('The user said:', userInput)

Enter some text: test
The user said: test


A full list of built-in functions and their description can be found at https://docs.python.org/3/library/functions.html.

### Random Numbers & Importing Modules

Sometimes we need random numbers, be it for a game or setting the initial values of a neural network. Python provides a module that has code for generating random numbers. A module is a file or collection of files of python code that we can use in our script using the `import` statement.  
Note: Every python file is a module and can be imported.

In [ ]:
import random

This statement loads the module random that is part of Python's standard library. `random` now refers to the module and we can use any function defined in random by prepending `random.` to the function's name, like for example, `random.randint(2, 4)`


For example, the function `random()` generates a float between 0 and 1, excluding 1.

In [ ]:
random.random()

0.7364712141640124

A full list of the random module's functions can be found at https://docs.python.org/3/library/random.html.  
For full documentation on the standard library and all the modules, see https://docs.python.org/3/library/index.html.

A note about random: Computers are deterministic and so `random` is not actually random. Random number generaters use a seed obtained from some arbitrary place (Python usually uses the current time for this) to generate numbers. Thus, if we use the same seed, we will get the same numbers every time.

In [ ]:
random.seed(42)
print(random.random())

random.seed(42)
print(random.random())

0.6394267984578837
0.6394267984578837


Notice how when we set the seed like in the code above, we get the same result every time. Therefore this is called pseudo-random numbers because they appear to be random but aren't.  
Note: Because of this predictability, never use random for anything involving encryption. Use `secrets` instead.

There are many functions in random, but one especially useful is `randint`.
It takes two arguments and will generate a random integer between the two, including both ends. 

In [ ]:
random.randint(2, 4)

4

On the note of importing again, we can also import just one function from a module if it is all we need using the keyword `from`.

In [ ]:
from random import randint

Now we can use `randint` on its own without specifying the module like `random.randint`

In [ ]:
randint(1, 10)

2

## Lists, Tuples, Dicts, Sets

### List

Often we need to keep track of many values, in which it would be easier to have a list instead of a dozen variables.  
Lists can track a variety of values and allows us to append, remove, and search the list.
Lists can also contain values of different types, although that is usually avoided.
  
Note: For people familiar with arrays from other languages, the list serves most of the time as the Python counterpart for the array.

In [ ]:
letters = ['a', 'b', 'c', 'd', 'e', 'f']

Values in a list are separated by commas and surrounded by square brackets.

To get the value at a specific index, use: variable\[ index \]

In [ ]:
letters[3]

'd'

D, however, is not the third letter of the alphabet. This is because in Python, as in all good programming languages, indexing starts at 0.  
So in the case of our letters list, index 0 is 'a'

In [ ]:
letters[0]

'a'

We can also set a value in a list.

In [ ]:
letters[0] = 'A'

In [ ]:
print(letters)

['A', 'b', 'c', 'd', 'e', 'f']


Lists also have a method that appends values to the end of the list. 
 
Note: a method is just like a function, but it is connected to an object like a list.

In [ ]:
letters.append('g')

Every time `letters.append('g')` is called, the letter g is added to the end of the list.

In [ ]:
print(letters)

['A', 'b', 'c', 'd', 'e', 'f', 'g']


We can also remove the 'g' again by calling `.remove('g')`.

In [ ]:
letters.remove('g')

list.remove removes only the first 'g', so if there were multiple occurrences of it, only the first one would be removed.

In [ ]:
print(letters)

['A', 'b', 'c', 'd', 'e', 'f']


We can also insert the word 'text' into our list at index 3 using `.insert(3, 'text').

In [ ]:
letters.insert(3, 'text')

In [ ]:
print(letters)

['A', 'b', 'c', 'text', 'd', 'e', 'f']


We can also remove it again using `.pop(3)`.

In [ ]:
letters.pop(3)

'text'

In [ ]:
print(letters)

['A', 'b', 'c', 'd', 'e', 'f']


We can also find an element in a list with `index`.

In [ ]:
letters.index('c')

2

The keyword `in` checks if the list contains an element, like:

In [ ]:
'b' in letters

True

There is also the inverse, `not in`.

In [ ]:
'b' not in letters

False

To find the length of a list, use `len()`. It is not a method of list, but a function that takes most types that contain multiple values.

In [ ]:
len(letters)

6

To create an empty list, use empty brackets: `[]`  
Other sequences can be turned into a list using `list()`.

In [ ]:
list( ('a', 3, 7.0) )

['a', 3, 7.0]

For a full list of the methods of a list, see https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types and https://docs.python.org/3/library/stdtypes.html#typesseq-common.  
Lists also have a method called `sort`, which is not mentioned in the linked pages and sorts the list.

### Tuple

Tuples are like a list, but they are immutable. That means they cannot be changed once created and tuples don't have methods like `append`, `remove`, etc. Tuples are enclosed with parenthesis instead of square brackets.

In [ ]:
("this", "is", "a", "tuple")

('this', 'is', 'a', 'tuple')

Though they may seem like lists but worse, this restriction gives them multiple use cases.  
One is that tuples are useful in assigning multiple variables at once, like:

In [ ]:
a, b = (2, 3)

print('a =', a)
print('b =', b)

a = 2
b = 3


This becomes especially useful when returning values from a function as it allows us to return more than one value.

In [ ]:
def squareRootBothSigns(x):
  root = x ** 0.5
  return -root, root

squareRootBothSigns(4)

(-2.0, 2.0)

This can be combined with the previous example to create an elegant way of returning multiple values from a function.

In [ ]:
minus, plus = squareRootBothSigns(4)

print('minus =', minus)
print('plus =', plus)

minus = -2.0
plus = 2.0


To create an empty tuple, use `()`.  

Note: As tuples use parenthesis, a single value in a tuple like `(3)` will be evaluated to `3`, as python thinks this is a 3 in parenthesis rather than a tuple. Add a comma at the end when there is only one value to specify that it is a tuple: `(3,)`

In [ ]:
type( (3) )

int

In [ ]:
type( (3,) )

tuple

`in`, `not in`, and `len()` also work with tuples.


Just like `list()` turns any sequence into a list, `tuple()` makes a tuple out of other sequence types.

In [ ]:
tuple([1,2,3,4])

(1, 2, 3, 4)

More methods can be found in https://docs.python.org/3/library/stdtypes.html#typesseq-common.

### Set

The set type is like the one found in discrete mathematics. It cannot contain duplicates and supports standard set operations like union and intersection and checking if a set is a subset of another. They also allow for efficient testing if an object is a member of a set using the keyword `in`.

In [ ]:
thisIsASet = {1, 2, 3, 4}

In [ ]:
thisIsASet.union({4, 5, 6})

{1, 2, 3, 4, 5, 6}

In [ ]:
thisIsASet.difference({2, 4, 6, 8, 10})

{1, 3}

In [ ]:
thisIsASet.issuperset({2, 4})

True

Sets do not support indexing so `thisIsASet[3]` will be an error.

Sets also support adding and removing elements like a list does using `add` and `remove`, where `add` is like `append`. 

To create an empty set, use `set()` with no arguments instead of `{}` as that is already used for the `dict` type, which will be covered next.

In [ ]:
type({})

dict

In [ ]:
type(set())

set

`in`, `not in`, and `len()` also work with sets. Using `in` with sets is much more efficient than with lists.

For converting to a set, use `set()`. This is also a way of filtering out repeated values.

In [ ]:
set([1, 3, 5, 7, 5, 3])

{1, 3, 5, 7}

See https://docs.python.org/3/library/stdtypes.html#set for all methods.

### Dictionaries

The dictionary, type `dict`, is a different data structure in that instead of being a list of values, it stores key-value pairs. Indexing is thus done by the key instead of by the numeric index.

In [ ]:
data = { 'a': 1, 'b': 2, 'c': 3 }

Each key-value pair is separated by a comma, and a colon separates the key from the value. Keys are on the left, values on the right.  
Indexing is done using the key.

In [ ]:
data['b']

2

Values can also be set with a key.

In [ ]:
data['b'] = -2

New key-value pairs can be added using the same syntax.

In [ ]:
data['d'] = 4

The `in` operator checks for the presence of keys, not values. This allows us to check if a key exists before indexing with it to avoid an error.

In [ ]:
'c' in data

True

In [ ]:
3 in data

False

In [ ]:
if 'e' in data:
  print(data['e'])
else:
  print('No value for e')

No value for e


`in`, `not in`, and `len()` also work with dictionaries.

Create empty dictionaries using `{}`.  
As dictionaries store pairs instead of values, using `dict()` to covert another list-like type to dict requires the given list to contain key-value pairs, like `[["student A", 100], ["student B", 95], ["student C", 90]]`. 

In [ ]:
dict([['first', 1], ['second', 2]])

{'first': 1, 'second': 2}

A list of methods can be found at https://docs.python.org/3/library/stdtypes.html#dict.

### Standard Functions For List-like Types

These functions will work on anything list-like, including `list`, `tuple`, `set`, and `dict`. In technical terms, anything that is an iterator or is iterable is accepted by these functions. `len()` may not work with iterators, however.

`sum()` will return the sum of the list.

In [ ]:
sum([1,2,3,4])

10

`max()` and `min()` return the largest and smallest value, respectively.

In [ ]:
max([1, 2, 3, 4])

4

In [ ]:
min([1, 2, 3, 4])

1

`len()` returns the length.

In [ ]:
len([1, 2, 3, 4])

4

`sorted()` returns it in order.

In [ ]:
sorted([4, 1, 3, 2])

[1, 2, 3, 4]

`reversed()` returns it in reverse order. However, it returns it as something known as an iterator, so use `list()` to turn that into a list. We could also use `tuple()` or `set()` in these cases.

In [ ]:
list(reversed([1, 2, 3, 4]))

[4, 3, 2, 1]


`filter()` filters a list based on the result of a function. It keeps it if the function returns true. There will be more explanation of comparisons and the `bool` type in the control flow segment.  
The first argument of filter is a function and the second the list.  
`filter` also returns an iterator, so use `list()` on it as well.

In [ ]:
def isPositive(x):
  return x > 0

list(filter(isPositive, [-2, -1, 0, 1, 2]))

[1, 2]

`map()` runs a function on every value of a given list and returns the result in a new list. Like filter, its first argument is a function and the second is the list. It also returns an iterator, so use `list()` on it.

In [ ]:
def square(x):
  return x * x

list(map(square, [1, 2, 3, 4]))

[1, 4, 9, 16]

## Control Flow

So far we can only do the same calculation every time, which doesn't make this better than a calculator. The next part is control flow, where we can make the program do different things based on the data we have.  


### The If Statement & Boolean Logic

If we want a certain piece of code to only run when a condition is met, we can use the if statement

In [ ]:
n = 3

if n > 0:
  print("Positive")

Positive


The syntax of the if statement is:
```
if condition:
  code
```
The condition can be anything and will be evaluated to either true or false.

True or False has its own type, which is bool. The constants for true is `True` and for false is `False` in python. 

In [ ]:
if True:
  print("This print statement always runs")

if False:
  print("This print statement never runs")

This print statement always runs


We can compare numbers as a condition, like `n > 0` in the first if example.
The valid comparisons are:
- greater than: `>`
- less than: `<`
- greater than or equal: `>=`
- less than or equal: `<=`
- equal to: `==`
- not equal: `!=`

Note: Notice that two equal signs `==` will check if the values are equal, while a single equal sign `=` will assign a variable.



In [ ]:
4 == 3

False

In [ ]:
9 <= 9

True

In [ ]:
9 < 9

False

Note: I mentioned before that floating-point numbers are not precise, so `==` might not consider numbers to be the same that we consider the same due to rounding errors.  
For example:

In [ ]:
0.1 + 0.2 == 0.3

False

In this case, it is because the addition of `0.1 + 0.2` produces rounding errors, making it `0.30000000000000004`. So it does not equal `0.3`.

In [ ]:
0.1 + 0.2

0.30000000000000004

The way to get around this is to check if the absolute value of the difference between the two numbers is a very small number, like:

In [ ]:
(0.1 + 0.2) - 0.3

5.551115123125783e-17

In [ ]:
abs( (0.1 + 0.2) - 0.3 ) < 0.000001

True

abs() is a built-in function in python that returns the absolute value.

booleans can also form more complex logic using the logical operators `and`, `or`, and `not`. For example, if we wanted to know if both x and y are positive, we could say:

In [ ]:
x = 4
y = 1

if x > 0 and y > 0:
  print("Both are postive")

Both are postive


More examples:

In [ ]:
not x < 0

True

In [ ]:
x < 0 or y < 0

False

In [ ]:
x > 0 and (y < 0 or y > 10)

False

`None`, when used as the condition of an if-statement, is considered False. We can explicitly check for `None` by checking if it equals None.

In [ ]:
if None:
  print("This will never run")

In [ ]:
a = None
if a == None:
  print("a is None")

a is None


### The If-Else Statement

If the condition in the if statement is false, the code inside the if statement is skipped. If we however wanted other code to run in the case of the condition being false, we can use `else`.

In [ ]:
n = -3

if n > 0:
  print("N is positive")
else:
  print("N is negative")

N is negative


The case above is not entirely correct as it would label 0 as a negative number. So we need another if statement inside of else to check for this third possible condition.

In [ ]:
n = 0

if n > 0:
  print("N is positive")
else:
  if n == 0:
    print("N is zero")
  else:
    print("N is negative")

N is zero


Because logic like the one above is quite common, there is a special keyword for it that combines the `else` and the inner `if`. The `elif` is like an `if` statement, but it is only run if the preceding `if` statement's condition is false.

In [ ]:
n = 0

if n > 0:
  print("N is positive")
elif n == 0:
  print("N is zero")
else:
  print("N is negative")

N is zero


`elif` can be repeated as many times as you like.

In [ ]:
n = 3

if n == 0:
  print("N is zero")
elif n == 1:
  print("N is one")
elif n == 2:
  print("N is two")
elif n == 3:
  print("N is three")
else:
  print("N is not 0, 1, 2, or 3")

N is three


We can combine comparisons with functions:

In [ ]:
def isEven(n):
  return n % 2 == 0

isEven(5)

False

As every even number can be divided by 2, the remainder for any even number divided by 2 must be zero.

We can then use functions as the condition in an `if` statement.

In [ ]:
x = 4
y = 10

if isEven(x) and isEven(y):
  print("X and Y are both even")

X and Y are both even


We can also use the `if` statement to write piecewise functions or recursive functions as seen in mathematics.

In [ ]:
def rampFunction(x):
  if x < 0:
    return 0
  else:
    return x

In [ ]:
def factorial(n):
  if n < 0:
    print("n must be greater than or equal to 0")
    
  elif n == 0 or n == 1:
    return 1
  else:
    return factorial(n - 1) * n


factorial(6)

720

### The While Loop

Sometimes we want to run code many times until some condition is met. For example, in the last example about if and else, I gave an example of a function that calculates factorial. If we instead wanted to write it without recursion, which is a function calling itself, we can use a `while` loop instead.

In [ ]:
def factorial(n):
  if n < 0:
    print("n must be greater than or equal to 0")
  

  result = 1
  while n > 0:
    result = result * n
    n = n - 1
  
  return result

factorial(6)

720

While loops are like the if statement, but they repeat for as long as the condition is true. Here it will repeat for as long as n is greater than 0 and on every iteration, it multiples the value to result. As n is decreased every round, we get that result = n * (n-1) * (n-2) * ... 1, which is n!.  
If the condition of the while loop is false initially, like if we called `factorial(0)`, then it is never run.

`while` is going to repeat for as long as its condition is true. This means it is **really easy** to create an infinite loop and your program will hang until it is forcefully stopped.  
On a command line like Terminal, Command Prompt, or IDLE, use Control-C to stop the program. Sometimes that doesn't work and you'll have to quit the application.  
In Jupyter Lab or Colab, there should be a stop button next to the cell that will stop the program.  
Note: KeyboardInterrupt Errors come from pressing Control-C or the stop button and don't mean that there is something wrong with the program. If the program never finishes, however, then you have an infinite loop, which is an issue.

Loops are especially useful when working with lists.

In [ ]:
def calculateSum(ls):
  index = 0
  result = 0
  while index < len(ls):
    result += ls[index]
    index += 1
  return result

calculateSum([1,2,3,4])

10

### The For Loop

A more pythonic way of writing the above function would be to use a for loop. For loops are like while loops, but run once for every element in a list.  
Because of this, never add or remove an element from a list that is being looped over. Additionally, for loops rarely cause infinite loops.

In [ ]:
def calculateSum(ls):
  result = 0
  for e in ls:
    result = result + e
  return result

calculateSum([1,2,3,4])

10

For every iteration of a for loop, the variable between `for` and `in` will be set to the current list element. I used `e` in my example but any variable could be used.

Usually for loops are the preferred way of working with lists in python as they are cleaner and in some cases, especially when working with the `list` type, more efficient.  
  
There are a few tools for working with for loops, like `reversed`, which reverses the order than a for loop iterates over the list.

In [ ]:
for e in reversed([1,2,3,4]):
  print(e)


4
3
2
1


Or if there are two lists to run over, `zip` will make a list of tuples, each containing one element of each given list.

In [ ]:
for a, b in zip(['a','b','c'], [1, 2, 3]):
  print(a, b)

a 1
b 2
c 3


When working with `dict`, the variable of the for loop will be set to a key. Then use indexing `[]` to get the corresponding value.

In [ ]:
data = {'a': 1, 'b': 2, 'c': 3}

for key in data:
  print('key =', key)
  print('value =', data[key])

key = a
value = 1
key = b
value = 2
key = c
value = 3


### Range

For loops can also be used to run a specific number of times, like in other languages, using `range()`.  
With one argument, `range(n)` will act as a list starting at 0 up to n-1, excluding n.

In [ ]:
counter = 0
for i in range(5):
  counter = counter + 1
  print(i)
  
print('The loop ran', counter, 'times')

0
1
2
3
4
The loop ran 5 times


With two arguments, `range(a, b)` will start at a and stop before b, excluding b.

In [ ]:
for i in range(5, 10):
  print(i)

5
6
7
8
9


With three arguments, `range(a, b, c)` will start at a and stop before b, excluding b and increment by c each iteration.

In [ ]:
for i in range(0, 10, 3):
  print(i)

0
3
6
9


### The Break Statement

Sometimes in a while or for loop, we want to stop suddenly when a condition is met. The break allows for this, ending the loop when run.

In [ ]:
ls = [1,2,3,4]
value = 3

foundValue = False
for e in ls:
  if e == value:
    foundValue = True
    break # found the value, so no need to continue

print("Found Value?", foundValue)


Found Value? True


### The Continue Statement

The continue statement is useful when in a loop certain code should be skipped. Once `continue` is run, it will skip the rest of the loop's code and continue to the next iteration.

In [ ]:

sumNotDivisibleBy7 = 0

for i in range(20):
  if i % 7 == 0:
    continue # 'Divisible by 7
  
  sumNotDivisibleBy7 = sumNotDivisibleBy7 + i

print(sumNotDivisibleBy7)

169


## Debugging

When programming, it is guaranteed that something will go wrong at some point. No one constantly writes perfect code on their first try. Flaws sneak into programs all the time, so learning to find them is just as important as learning to program.  
I'm gonna cover how I would approach a piece of dysfunctional code to find the mistake I made.  
I'm gonna be using the print statement. There are fancier debugging tools, but for most cases in python, print will do the job.

In [ ]:
def mode(ls):
	d = {}  # make a dictionary that counts how often each value appears
	for value in ls:
		if value in d:
			d[value] += 1  # seen it before, increase by one
		else:
			d[value] = 1    # never seen this value

	maxValue = ls[0]   # now find the most occuring value
	for key in ls:
		if ls[key] > ls[maxValue]:   # if value key occurs more often than maxValue, it is the new maxValue
			maxValue = key
	return maxValue

mode([1,2,2,3])

3

This is a piece of code I intended to put a small bug into to show how debugging works, but in the process of writing it, I managed to do it accidentally.  
The function returns 3 as the mode, but 2 is the one that occurs most often.  

First, we have to narrow down where the problem is.  
In the first part, we track in a `dict` how often each item occurs in the list. We do that by having the keys in our `dict` d be the elements in ls and the corresponding value being the number of times we have seen it. When the element is not in d, we haven't seen it yet so we set the times we've seen it to once as this is our first time.  
If we have seen the element (it exists as a key in d), we increment it by 1.  
  
So we add `print(d)` to verify if this for loop is correct. If it is, we should see `{1: 1, 2: 2, 3: 1}`.

In [ ]:
def mode(ls):
	d = {}
	for value in ls:
		if value in d:
			d[value] += 1
		else:
			d[value] = 1
	print(d)  # checks if the first part when right
	maxValue = ls[0]
	for key in ls:
		if ls[key] > ls[maxValue]:
			maxValue = key
	return maxValue

mode([1, 2, 2, 3])

{1: 1, 2: 2, 3: 1}


3

We do get what we expected, so the error must be in the second part of our code.  
So I added another print statement to check on what is happening in the second for loop.

In [ ]:
def mode(ls):
	d = {}
	for value in ls:
		if value in d:
			d[value] += 1
		else:
			d[value] = 1
	print(d)
	maxValue = ls[0]
	for key in ls:
		print(key, ls[key])  # check what’s being compared
		if ls[key] > ls[maxValue]:
			maxValue = key
	return maxValue

mode([1, 2, 2, 3])

{1: 1, 2: 2, 3: 1}
1 2
2 2
2 2
3 3


3

Then we analyze the result and check if it is what we expected.
> {1: 1, 2: 2, 3: 1}  
> 1 2 <- makes no sense. The key-value pair in d is 1: 1   
> 2 2  
> 2 2  
> 3 3 <- makes no sense. The key-value pair in d is 3: 1  
> 3  
  
We are comparing how many times each value appears, so `1 2` and `3 3` both make no sense.  
While looking over the section, it should appear that although we want to compare data from d, we are getting it from ls, which is the wrong variable.


In [ ]:
def mode(ls):
	d = {}
	for value in ls:
		if value in d:
			d[value] += 1
		else:
			d[value] = 1
	maxValue = ls[0]
	for key in d: # replaced ls with d
		if d[key] > d[maxValue]: # replaced ls with d
			maxValue = key
	return maxValue

mode([1, 2, 2, 3])

2

Replacing ls with d in the second for loop fixed the problem and we are getting a result of 2, which is what we expected for the given input.